src/SLCore/State/AliveConnectionTracker.cs

```diff
@@ -0,0 +1,135 @@
+    
+    [ImportingConstructor]
+    public AliveConnectionTracker(ISLCoreServiceProvider serviceProvider,
+        ISolutionBindingRepository solutionBindingRepository,
+        IConnectionIdHelper connectionIdHelper,
+        IAsyncLockFactory asyncLockFactory,
+        IThreadHandling threadHandling)
+    {
+        this.serviceProvider = serviceProvider;
+        this.solutionBindingRepository = solutionBindingRepository;
+        this.threadHandling = threadHandling;
+        asyncLock = asyncLockFactory.Create();
+        this.connectionIdHelper = connectionIdHelper;
+
+        solutionBindingRepository.BindingUpdated += BindingUpdateHandler;
+    }
+
+    public async Task RefreshConnectionListAsync()
+    {
+        threadHandling.ThrowIfOnUIThread();
+
+        if (!serviceProvider.TryGetTransientService(out IConnectionConfigurationSLCoreService connectionConfigurationService))
+        {
+            throw new InvalidOperationException(Strings.ServiceProviderNotInitialized);
+        }
+
+        using (await asyncLock.AcquireAsync())
+        {
+            var serverConnections = GetUniqueConnections(solutionBindingRepository.List());
+
+            await connectionConfigurationService.DidUpdateConnectionsAsync(new DidUpdateConnectionsParams(
+                serverConnections.Values.OfType<SonarQubeConnectionConfigurationDto>().ToList(),
+                serverConnections.Values.OfType<SonarCloudConnectionConfigurationDto>().ToList()));
+
+            foreach (var connectionId in serverConnections.Keys)
+            {
+                // we don't manage connections as separate entities and we don't know when credentials actually change
+                await connectionConfigurationService.DidChangeCredentialsAsync(
+                    new DidChangeCredentialsParams(connectionId));
+            }
+        }
+    }
+
+    private void BindingUpdateHandler(object sender, EventArgs arg)
+    {
+        threadHandling.RunOnBackgroundThread(async () =>
+        {
+            await RefreshConnectionListAsync();
+
+            return 0;
+        }).Forget();
+    }
+
+    private Dictionary<string, ServerConnectionConfiguration> GetUniqueConnections(IEnumerable<BoundSonarQubeProject> bindings)
+    {
+        var connections = new Dictionary<string, ServerConnectionConfiguration>();
+
+        foreach (var binding in bindings)
+        {
+            var serverUri = binding.ServerUri;
+            var organization = binding.Organization?.Key;
+            var connectionId = connectionIdHelper.GetConnectionIdFromUri(serverUri, organization);
+
+            if (serverUri == ConnectionIdHelper.SonarCloudUri)
+            {
+                connections[connectionId] = new SonarCloudConnectionConfigurationDto(connectionId, true, organization);
+            }
+            else
+            {
+                connections[connectionId] = new SonarQubeConnectionConfigurationDto(connectionId, true, serverUri.ToString());
+            }
```

Comment:
This should be an assignment from ternary, to structurally ensure that the assigned thing is always the same
```
connections[connectionId] =serverUri == ConnectionIdHelper.SonarCloudUri
    ? new SonarCloudConn...
    : new SonarQubeConn...

```